### PR TITLE
Use Kotlin context element in agent instrumentation

### LINKED ISSUE #1411: KotlinProbeInstrumentation does precisely what gRPC Javadoc says not to
[CoroutineContextWrapper.tracingSuspend()](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/kotlin-coroutines/src/main/java/io/opentelemetry/javaagent/instrumentation/kotlincoroutines/KotlinProbeInstrumentation.java#L194):
```
public void tracingSuspend() {
    myTracingContext = Context.current();
    myTracingContext.detach(prevTracingContext);
}
```
Apart from saving the current context, the above code is equivalent to `Context.current().detach()`, which gRPC [Context.detach()](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html#detach-io.grpc.Context-) asks not to do:
> **Never** use `Context.current().detach()`, as this will compromise this error-detecting mechanism.

On an aside, I noticed that gRPC context is used everywhere for propagation. This is very odd; OTEL ought to make their own. Most people wouldn't expect or want a tracing library to quietly pull in gRPC and its dependencies (Guava etc.).

### PR DESCRIPTION
**Pull request by anuraaga**
Follow-on to #1617 

@johnbley @iNikem mentioned he would chat with you about the approach of using context element, including the notion that `makeCurrent` in kotlin code is illegal and shouldn't really be supported by us. Figured having this in front of your eyes will help the discussion :)

Fixes #1411

### CODE REVIEW

File: instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt [hunk 1]
```diff
@@ -145,35 +147,35 @@ class KotlinCoroutineTests(private val dispatcher: CoroutineDispatcher) {
   suspend fun a(iter: Long) {
     var span = tracer.spanBuilder("a").startSpan()
     span.setAttribute("iter", iter)
-    var scope = span.makeCurrent()
-    delay(10)
-    a2(iter)
-    scope.close()
+    withContext(span.asContextElement()) {
```
Comment (anuraaga) [inline instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: These changes are the key to the proposal. My theory is `span.makeCurrent` is illegal code - the only reason it works here is because the agent is present, but if the agent was not present, resumed coroutines could see completely unrelated contexts, which is a disaster. The idea is that our agent is here to enable tracing, not to allow illegal code. How to "codify" that it's illegal is a tricky point though but something we will want to think about. Current idea is that https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/testing/src/main/java/io/opentelemetry/sdk/testing/context/StrictContextStorage.java#L99 can also check the stacktrace to see if it is being called from a coroutine, and not from the context element. Dunno if it'll work but if it does, I think we end up in a nice place :)

Comment (anuraaga) [reply instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: Also kudos to @adriancole for giving the idea of stack-walking as a potential mechanism for checking

Comment (johnbley) [reply instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: Does this require application authors who use auto-instrumentation and simply use (e.g.) kotlin suspend functions in their logic (without any manual instrumentation) to explicitly use (kotlin) `withContext` to pass our (otel) context?  That is, if app logic is written in kotlin and taking advantage of auto-instrumentation of inbound and outbound connections, will this still work without the app author having to take extra steps?

Comment (anuraaga) [reply instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: Don't think so - the agent with this code is making sure to automatically propagate the parent at the time the coroutine is launched into its context, so that parent is always available. If all they're doing is using libraries which are auto-instrumented by the agent, it should work fine without calling `withContext`. It's only when doing manual instrumentation within Kotlin that they need to use `withContext` instead of `makeCurrent`. How does that sound?

Comment (trask) [reply instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: I can see the confusion of this particular test, since it's a javaagent test, but also uses the library instrumentation directly.

do you think we need the nested spans to validate `launchConcurrentSuspendFunctions`? if not, we could remove `a2` and `b2` and not create any nested scopes at all, and not need to use the library instrumentation in the test.

if you think the nested scopes is a good test showing interop, maybe we can move it out into a separate test class `...InteropTest`.

**Review by trask — APPROVED** [review]
nice library instrumentation breakout 👍 and great point about the problem of `makeCurrent` under kotlin coroutines 😬

File: instrumentation/kotlin-coroutines/library/src/main/kotlin/io/opentelemetry/instrumentation/kotlincoroutines/ContextExtensions.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import io.opentelemetry.api.trace.Span
+import io.opentelemetry.context.Context
+import io.opentelemetry.javaagent.instrumentation.api.Java8BytecodeBridge
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Returns a [CoroutineContext] which will make this [Context] current when resuming a coroutine
+ * and restores the previous [Context] on suspension.
+ */
+fun Context.asContextElement(): CoroutineContext {
```
Comment (anuraaga) [inline instrumentation/kotlin-coroutines/library/src/main/kotlin/io/opentelemetry/instrumentation/kotlincoroutines/ContextExtensions.kt:17]: By the way, I'm wondering whether these should be in the SDK instead of instrumentation. Mainly since I was imagining adding to the javadoc for `makeCurrent` to not use it in Kotlin coroutines and use this instead. But that doesn't make as much sense when crossing repos.

Comment (trask) [reply instrumentation/kotlin-coroutines/library/src/main/kotlin/io/opentelemetry/instrumentation/kotlincoroutines/ContextExtensions.kt:17]: language-specific extensions to the API, e.g. `opentelemetry-extension-kotlin`? I like it

Comment (anuraaga) [reply instrumentation/kotlin-coroutines/library/src/main/kotlin/io/opentelemetry/instrumentation/kotlincoroutines/ContextExtensions.kt:17]: Sent https://github.com/open-telemetry/opentelemetry-java/pull/2066 for the library aspect of this PR

File: instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt (hunk 1 above)
Comment (johnbley) [reply instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: The nesting spans are crucial to the test - without any form of instrumentation, we were previously reporting nonsensical traces like "a causes b causes a2" and "b2 causes a".

Comment (johnbley) [reply instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: > . It's only when doing manual instrumentation within Kotlin that they need to use `withContext` instead of `makeCurrent`. How does that sound?

That sounds great; I have no problem making kotlin instrumentation authors work a little to preserve (otel) context.  I actually originally tried something like this (using the kotlin coroutine context system more "naturally) but was unable to make it work.  I'm glad you were able to figure it out!  l'm going to try to find some time to manually validate with a "real" kotlin app and see how this change works out in practice.

File: instrumentation/kotlin-coroutines/library/src/main/kotlin/io/opentelemetry/instrumentation/kotlincoroutines/ContextExtensions.kt (hunk 2 above)
Comment (johnbley) [reply instrumentation/kotlin-coroutines/library/src/main/kotlin/io/opentelemetry/instrumentation/kotlincoroutines/ContextExtensions.kt:17]: I agree - this is a great approach.

**Review by johnbley — APPROVED** [review]

File: instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt (hunk 1 above)
Comment (holyjak) [reply instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt:150]: @johnbley #1620 links to a real, though simple kotlin app that exhibits problems with span propagation

**Comment by holyjak** [comment]
Great work! 
I would like to use OT in a Kotlin app now(ish). Is there a workaround I can apply or is it best to wait a few days for this to be merged and then build the agent from `master`? Thank you!

**Comment by iNikem** [comment]
@anuraaga What are your plans for making this real PR?

**Comment by anuraaga** [comment]
@iNikem Going to do it after updating to snapshots :)

**Comment by anuraaga** [comment]
Updated to use the SDK's kotlin extension

**Review by mateuszrzeszutek — APPROVED** [review]
